lasagne/layers/shape.py

```diff
@@ -26,6 +27,78 @@ def get_output_for(self, input, *args, **kwargs):
 flatten = FlattenLayer # shortcut
 
 
+class ReshapeLayer(Layer):
+    """
+    A layer reshaping its input tensor to another tensor of the same total
+    number of elements.
+
+    :parameters:
+        - input_layer : `Layer` instance
+            The layer from which this layer will obtain its input
+
+        - shape : tuple
+            The target shape. Any of its elements can be `None`, denoting to
+            retain the size of the input shape for this dimension. At most one
+            element can be `-1`, denoting to infer the size for this dimension
+            to match the total number of elements of the input shape. Any
+            remaining elements must be positive integers.
+
+    :usage:
+        >>> from lasagne.layers import InputLayer, ReshapeLayer
+        >>> l_in = InputLayer((100, 20))
+        >>> l1 = ReshapeLayer(l_in, (None, 2, 10))
+        >>> l1.get_output_shape()
+        (100, 2, 10)
+        >>> l2 = ReshapeLayer(l_in, (None, 1, 2, 5, -1))
+        >>> l2.get_output_shape()
+        (100, 1, 2, 5, 2)
+
+    :note:
+        The tensor elements will be fetched and placed in C-like order. That
+        is, reshaping `[1,2,3,4,5,6]` to shape `(2,3)` will result in a matrix
+        `[[1,2,3],[4,5,6]]`, not in `[[1,3,5],[2,4,6]]` (Fortran-like order),
+        regardless of the memory layout of the input tensor. For C-contiguous
+        input, reshaping is cheap, for others it may require copying the data.
+    """
+
+    def __init__(self, input_layer, shape):
+        super(ReshapeLayer, self).__init__(input_layer)
+        shape = tuple(shape)
+        if not all(s is None or isinstance(s, int) for s in shape):
+            raise ValueError("`shape` must be a tuple of int and/or None")
+        if any(s is not None and (s == 0 or s < -1) for s in shape):
+            raise ValueError("`shape` integers must be positive or -1")
+        if sum(s == -1 for s in shape) > 1:
+            raise ValueError("`shape` cannot contain multiple -1")
+        self.shape = shape
+
+    def get_output_shape_for(self, input_shape, *args, **kwargs):
+        # First, replace all `None` with the corresponding input dimension
+        output_shape = list(self.shape)
+        for dim, o in enumerate(output_shape):
+            if o is None:
+                output_shape[dim] = input_shape[dim]
+        # Secondly, infer value for -1 if needed
+        if -1 in output_shape:
+            dim = output_shape.index(-1)
+            output_shape[dim] = np.prod(input_shape) // -np.prod(output_shape)
```

Comment:
If any `input_shape` dimension is `None`, this will fail. If all such dimensions are copied over to `output_shape`, we could still correctly infer the value for `-1`, though.
Example: for `input_shape = (None, 100)`, `self.shape = (None, -1, 20)`, we can infer `output_shape = (None, 5, 20)`.